### ISSUE TITLE
LSAN not aborting in persistent mode

### ISSUE THREAD

**Issue by MegaManSec** [AFLplusplus/AFLplusplus#825]
It seems that leaksanitizer (under ASAN) does not work in AFL++'s persistence mode.

The following C code produces a memory leak:
```
#include <stdio.h>
#include <string.h> 
#include <stdlib.h>
#include <unistd.h>
//#include <sanitizer/lsan_interface.h>

__AFL_FUZZ_INIT();
int main() {
__AFL_INIT();
unsigned char *b = __AFL_FUZZ_TESTCASE_BUF;
while(__AFL_LOOP(1000)) {
   size_t len = __AFL_FUZZ_TESTCASE_LEN;
   if(len >= 100 || len == 0) continue;
   char *leak = malloc(len+1);
   memcpy(leak,b,len);
   leak[len] = '\0';
   fprintf(stderr,"%s\n", leak);
   fflush(stderr);
}
}
```

`export ASAN_OPTIONS='abort_on_error=1:detect_leaks=1:symbolize=0' ; AFL_USE_ASAN=1 afl-clang-lto try.c ; printf "AAAAAA" | ./a.out`
produces the correct report:
```
afl-cc ++3.11a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-LTO-PCGUARD
afl-llvm-lto++3.11a by Marc "vanHauser" Heuse <mh@mh-sec.de>
Info: Found constructor function asan.module_ctor with prio 1, we will not instrument this, putting it into a block list.
[+] Instrumented 21 locations with no collisions (on average 0 collisions would be in afl-gcc/vanilla AFL) (non-hardened, ASAN mode).
AAAAAA

=================================================================
==1536867==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 7 byte(s) in 1 object(s) allocated from:
    #0 0x2ca6dd  (/tmp/a.out+0x2ca6dd)
    #1 0x2feabd  (/tmp/a.out+0x2feabd)

SUMMARY: AddressSanitizer: 7 byte(s) leaked in 1 allocation(s).
Aborted
```

We then run the same program in afl-fuzz:
`mkdir test/ ; printf "AAAA" > test/1`
`afl-fuzz -i test/ -o test-out/ ./a.out`

but there are no crashes.

I tried to run add a call the __lsan_do_leak_check(); at the end of the loop, but this does not cause a crash either.

**Comment by MegaManSec** [comment]
In fact, it seems that it does not work in any mode:
```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
int main(int argc, char** argv) {
	char *b = argv[1];
	int len = strlen(b);
	char *leak = malloc(len+1);
	memcpy(leak,b,len);
	leak[len] = '\0';
	fprintf(stderr,"%s\n", leak);
	leak = NULL;
	fflush(stderr);
}
```
with ASAN_OPTIONS='abort_on_error=1:detect_leaks=1:symbolize=0'
aborts when running ./a.out hello

but not with `AFL_USE_ASAN=1 afl-fuzz -i test/ -o test-out/ ./a.out @@`

**Comment by vanhauser-thc** [comment]
and where is the bug? This is how it is supposed to work :)
afl-fuzz sets default ASAN_OPTIONS that do not abort on leaks, because this would trigger tons of crashes constantly for most software, and that is not what people are looking for.

for people specifically looking for leaks you can set ASAN_OPTIONS and then it works.

BUT - it is a bad approach to finding leaks IMHO.
it is easier to just fuzz normally. and then use the queue entries on an ASAN compiled target with the proper ASAN_OPTIONS set to see if any queue entries trigger leaks.
this is not a memory corruption like thing where asan would need to be present for every fuzz attempt. missing a leak by just going through the queue is unlikely.

**Comment by MegaManSec** [comment]
Yes, below the code I mentioned I used `export ASAN_OPTIONS='abort_on_error=1:detect_leaks=1:symbolize=0'` but no crash still

**Comment by vanhauser-thc** [comment]
I dont really know how the leak sanitizer works.
I know that leaks are reported on program exit which would be at the end of an __AFL_LOOP(n), so not what you want.
I would have thought that immediate leaks would be detected in real time, but apparently that is not the case.
I do not know if a leak also results in a core (I do not see how this would help leak analysis).
that is something you should find out :)

**Comment by MegaManSec** [comment]
They preform at exit unless `__lsan_do_leak_check()` is used, which will perform the leak check wherever it is run. It results in an SIGSTOP if a leak is found:
```
$ clang -fsanitize=address,leak a.c
$ ./a.out ssssadasda324r233
ssssadasda324r233

=================================================================
==23202==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 18 byte(s) in 1 object(s) allocated from:
    #0 0x407808 in malloc (/tmp/a.out+0x407808)
    #1 0x4289d4 in main (/tmp/a.out+0x4289d4)
    #2 0x7ffff6e22bf6 in __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:310

SUMMARY: LeakSanitizer: 18 byte(s) leaked in 1 allocation(s).
jrogers@10:/tmp$ echo $?
23
```



I don't know whether this is a bug in my version of clang (unlikely?) but I have now noticed that using __lsan_do_leak_check() only works when I strictly pass -fsanitize=leak to the compiler:
```
jrogers@10:/tmp$ clang -fsanitize=address a.c
jrogers@10:/tmp$ ASAN_OPTIONS=detect_leaks=1 a.out 11111
bash: a.out: command not found
jrogers@10:/tmp$ ASAN_OPTIONS=detect_leaks=1 a.out 11111^C
jrogers@10:/tmp$ clang -fsanitize=address a.c
jrogers@10:/tmp$ ASAN_OPTIONS=detect_leaks=1 ./a.out 11111
11111
jrogers@10:/tmp$ clang -fsanitize=address,leak a.c
jrogers@10:/tmp$ ASAN_OPTIONS=detect_leaks=1 ./a.out 11111
11111
jrogers@10:/tmp$ clang -fsanitize=leak a.c
jrogers@10:/tmp$ ASAN_OPTIONS=detect_leaks=1 ./a.out 11111
11111

=================================================================
==30960==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 6 byte(s) in 1 object(s) allocated from:
    #0 0x407808 in malloc (/tmp/a.out+0x407808)
    #1 0x4289d4 in main (/tmp/a.out+0x4289d4)
    #2 0x7ffff6e22bf6 in __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:310

SUMMARY: LeakSanitizer: 6 byte(s) leaked in 1 allocation(s).
```

**Comment by vanhauser-thc** [comment]
if you want you can write a section in the docs/FAQ,md on how to perform leak hunting

**Comment by MegaManSec** [comment]
Alright so things discovered:

1. compiling with AFL_USE_ASAN (-fsanitize=address) will cause the program to show leaks when running the program by itself.
2. compiling with AFL_USE_ASAN and using afl-fuzz will NOT show the leaks for some reason at all.
3. editing src/afl-cc.c to add a new option, AFL_USE_LSAN to pass -fsanitize=leak will make the program show leaks when running and when using afl-fuzz. No ASAN_OPTIONS needed.
4. LSAN _exit()s on finding a leak instead of abort. 

I know that msan does not abort on error either so I assume somehow afl is reading the exit code. Where is this handled? i will write the code and submit a pull request fyi

**Comment by vanhauser-thc** [comment]
for 3.: you would normally do CFLAGS=-fsanitize=leak ... but you can send a PR for the USE_LSAN if you want

for the detection: LSAN_OPTIONS="exitcode=32"  <= 32 is the default.

for the implementation check out:
```
src/afl-forkserver.c:1214:          (fsrv->uses_asan && WEXITSTATUS(fsrv->child_status) == MSAN_ERROR)
```

**Comment by MegaManSec** [comment]
I will get this done.. soon (just a bit busy right now)

**Comment by vanhauser-thc** [comment]
no hurry. 3.20c release is further away :)

**Comment by MegaManSec** [comment]
does `__AFL_LEAK_CHECK();` seem like a reasonable command to run the leak checker?

**Comment by vanhauser-thc** [comment]
Imho yes

**Comment by MegaManSec** [comment]
Closed by #855
